glom/core.py

```diff
@@ -1334,32 +1334,62 @@ def __repr__(self):
 
     def __getstate__(self):
         t_path = _T_PATHS[self]
-        return tuple(('T' if t_path[0] is T else 'S',) + t_path[1:])
+        return tuple(({T: 'T', S: 'S', A: 'A'}[t_path[0]],) + t_path[1:])
 
     def __setstate__(self, state):
-        _T_PATHS[self] = (T if state[0] == 'T' else S,) + state[1:]
+        _T_PATHS[self] = ({'T': T, 'S': S, 'A': A}[state[0]],) + state[1:]
 
 
 _T_PATHS = weakref.WeakKeyDictionary()
 
 
 def _t_child(parent, operation, arg):
     t = TType()
-    _T_PATHS[t] = _T_PATHS[parent] + (operation, arg)
+    base = _T_PATHS[parent]
+    if base[0] is A and operation not in ('.', '[', 'P'):
+        # whitelist rather than blacklist assignment friendly operations
+        # TODO: error type?
+        raise BadSpec("operation not allowed on A assignment path")
+    _T_PATHS[t] = base + (operation, arg)
     return t
 
 
+def _s_first_magic(scope, key, _t):
+    """
+    enable S.a to do S['a'] or S['a'].val as a special
+    case for accessing user defined string variables
+    """
+    err = None
+    try:
+        cur = scope[key]
+    except (KeyError, IndexError) as e:
+        err = NameError("name '%s' not defined in glom scope" % key)
```

Comment:
Just occurred to me, seems like this should be an attribute error or something. Just thinking about how this should maybe be a PathAccessError and _t_eval which calls this doesn't turn it into one.

Reply:
Compare when you assign to something that doesn't exist:

```
>>> glom({'x': 123}, ('x', A.b.c, S.b.c))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mahmoud/projects/glom/glom/core.py", line 1944, in glom
    raise err
glom.core.wrap(NameError): error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: {'x': 123}
 - Spec: ('x', A.b.c, S.b.c)
 - Spec: 'x'
 - Target: 123
 - Spec: A.b.c
NameError: name 'b' not defined in glom scope
```
vs when you access something that doesn't exist:
```
>>> glom({'x': 123}, ('x', A.b, S.b.c))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mahmoud/projects/glom/glom/core.py", line 1944, in glom
    raise err
glom.core.PathAccessError: error raised while processing, details below.
 Target-spec trace (most recent last):
 - Target: {'x': 123}
 - Spec: ('x', A.b, S.b.c)
 - Spec: 'x'
 - Target: 123
 - Spec: A.b
 - Spec: S.b.c
glom.core.PathAccessError: could not access 'c', part 1 of T.b.c, got error: AttributeError("'int' object has no attribute 'c'")
```

Reply:
I copied from here:

```python
>>> a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'a' is not defined
```

I guess this comes down to -- are scopes more like objects (in which case AttributeError), or more like actual python scopes (in which case NameError)

I think there was a whole conversation about "syntax" errors vs "runtime errors".  I can't remember the details of the earlier convo, but looking at it now it seems that `S.a.b` and `S.a` should have approximately the same behavior (so undefined variable is more of a "runtime error" that `Or()` or `Coalesce()` can work around)

Reply:
+1 PathAccessError; I don't feel super strongly about AttributeError vs NameError